Migrate components to standalone

hudson-newey opened the pull request:
# Migrate components to standalone

This pull request is a draft.

As with most migration scripts, the standalone migration script gave me a good start, but there is still a lot of work to do.

I have opened a PR so that I can more clearly see what the migration script modified, and what work I have left to do.

---

This branch is forked from the `compiler-control-flow` branch because migrating directive based control flow (e.g. `*ngIf`, `*ngFor`, etc...) removed `CommonModule` as a dependency.

To reduce the merge conflicts that are bound to happen in the future, I have reduced the amount of merge work that I have to do in the future by basing this PR off of the `compiler-control-flow` branch.

## Changes

- **Important!** Adds Prettier as a dev dependency and standardizes default formatter to prettier

## Issues

Fixes: #2183 

## Final Checklist

- [x] Assign reviewers if you have permission
- [x] Assign labels if you have permission
- [x] Link issues related to PR
- [ ] Ensure project linter is not producing any warnings (`npm run lint`)
- [x] Ensure build is passing on all browsers (`npm run test:all`)
- [x] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

hudson-newey committed:
Migrate to new control flow syntax

hudson-newey committed:
Remove some whitespace churn

hudson-newey committed:
Migrate components to standalone

Fixes: #2183

hudson-newey committed:
Run remove components from modules migration

hudson-newey committed:
Fix failed migration

hudson-newey committed:
Some linting fixes

hudson-newey committed:
Fix some missed control flow directives

hudson-newey committed:
Staging to add automatic formatting

Whenever these migrations run, I get a bunch of eslint errors that could
easily be fixed by a formatter.

The most recent migration produced hundreds of linter errors that could
be fixed by automatic formatting.

Therefore, I'm standardising our formatting rules

hudson-newey committed:
Add prettier as a dev dependency

hudson-newey committed:
Run formatter over project

(This created a lot of churn. We might want to exclude this commit from
the git blame)

hudson-newey committed:
Revert "Run formatter over project"

This reverts commit 71af9eedad95c5e15b8e9a114f32f9f41a6919c9.

hudson-newey committed:
Re-run formatter with improved config

hudson-newey committed:
Revert "Re-run formatter with improved config"

This reverts commit 121b10bcae0195b481226bcd208ceaa3be8b8811.

hudson-newey committed:
Revert prettier changes in this PR

I want to keep these PRs as isolated as possible so that during review,
there is no overlap and formatting churn is kept in one commit with one
PR.

Formatting is therefore outside the scope of this PR.

hudson-newey committed:
Fix migration script errors

hudson-newey committed:
Migrate a lot of shared components

hudson-newey committed:
Fix build to see bundle size

I want to use the CI to see how much the bundle size has reduced by
since migrating some components to standalone.

(yes I know I can do it locally, but comparing CI builds is a better
test because it uses an immutable environment)

hudson-newey committed:
Remove pipes module

hudson-newey committed:
Remove datatable directives module

hudson-newey committed:
Remove items module

hudson-newey committed:
Remove directives module

hudson-newey committed:
Greatly reduce shared module

I think I can safely remove this module. However, I have to tripple
check

hudson-newey committed:
Remove shared module

hudson-newey committed:
Fix some imports

hudson-newey committed:
Remove shared module from tests

This means that all tests now compile, but a lot fail because they do
not have component declarations

hudson-newey committed:
Fix a lot of broken tests

hudson-newey committed:
Complete all easy test fixes

hudson-newey committed:
Remove unused BrowserAnimationModule

hudson-newey committed:
Fix some harder tests

hudson-newey committed:
All component > standalone test migration fixes

hudson-newey committed:
Fix formly modules

Formly doesn't support standalone (until version 7)

Therefore, I have to use a module system.

I also consolidated the formly modules so that there is one provider for
the formly module.

hudson-newey committed:
Add BrowserModule to formly

hudson-newey committed:
All semi-difficult tests fixed

hudson-newey committed:
Remove (most) RoutingTestingModule

hudson-newey committed:
Fully remove RouterTestingModule

hudson-newey committed:
Remove all exported components from modules

hudson-newey committed:
Fix annotation search form tests

hudson-newey committed:
Fix new region tests

atruskie commented on .prettierignore:
```diff
@@ -1,7 +1,11 @@
 src/app/services/baw-api/ServiceTokens.ts
+
+# we ignore .html files because prettier doesn't play nice with compiler
+# control flow (e.g. @if, @else, and @for)
```
open issue on prettier? cross link?

hudson-newey replied:
Looks like there is a `--parser angular` option that I should be using.

And it formats angulars new control flow syntax correctly

hudson-newey replied:
Fixed in: [`6a42613` (#2270)](https://github.com/QutEcoacoustics/workbench-client/pull/2270/commits/6a426138fe3cce7650554621412d1f506723c8bb)

Note: this explicitly adds prettier as a dev dependency.

atruskie reviewed (approved):
This is too large for an effective review.

Talk to me off line about those pain points and walk me through the problems

hudson-newey committed:
Refactor out page title strategy service

hudson-newey committed:
Remove app.module

hudson-newey committed:
Remove PageComponent modules

hudson-newey committed:
Refactor out ApplicationConfig

hudson-newey committed:
Add helper for indexed strong routes

hudson-newey committed:
Refactor page component definitions out to .pages

hudson-newey committed:
Remove routing module

hudson-newey committed:
Remvoe a lot of modules

hudson-newey committed:
Move app.config

hudson-newey committed:
Remove config module

[skip ci]

hudson-newey committed:
Remove baw api module

hudson-newey committed:
Remvoe caching module

[skip ci]

hudson-newey committed:
Migrate admin pages

hudson-newey committed:
Use standalone bootstrapping in ssr

hudson-newey committed:
Fix compiler errors for tests

hudson-newey committed:
provide ConfigService through provide-configMock

hudson-newey committed:
Provide ng-http-caching through baw-apiMock

This is not new, I just forgot to migrate it when I migrated away
from ng-modules.

hudson-newey committed:
Fix provider errors in tests

hudson-newey committed:
Add prettier as dev dependency to workspace

hudson-newey committed:
Import NgTemplateOutlet in toast provider

hudson-newey committed:
Fix linting errors

hudson-newey committed:
Casing change

hudson-newey committed:
Revert casing change

I needed to do these two commits because file names on windows are
case-insensitive

hudson-newey committed:
Small formatting changes to force CI run

hudson-newey committed:
Do not use angular parser for typescript files

hudson-newey committed:
Mock api services in some tests

hudson-newey committed:
Re-enable some disabled tests

I disabled these tests during migration because they were failing, but I
should really get them working

hudson-newey committed:
Fix failing tests

Some tests were previously giving an error "Expression changed after
checked" error.

This only occured when the AnnotationSearchFormComponent and
AnnotationSearchPageComponent tests were run together. Running the tests
individually did not have the same error.
This indicates that we have some leakage between test cases.

hudson-newey committed:
DO NOT provide HttpClientTesting in dev server

For some reason I made this mistake and it destroyed the dev server
because no requests could be made.
It was also really hard to debug because there were no error messages.

hudson-newey committed:
Revert eager bootstrapping

I don't fully understand why we wait for the DOMContentLoaded event to
bootstrap Angular. Especially since it is not the recommended behavior
by the Angular team.

I want to fully understand why we delay bootstrapping to ensure that I
don't accidently introduce an undocumented bug that previously existed.
